x-pack/plugins/fleet/server/routes/security.ts

```diff
@@ -165,22 +93,199 @@ export async function getAuthzFromRequest(req: KibanaRequest): Promise<FleetAuth
   });
 }
 
-export type RouterWrapper = <T extends RequestHandlerContext>(route: IRouter<T>) => IRouter<T>;
+function hasRequiredFleetAuthzPrivilege(
+  authz: FleetAuthz,
+  {
+    fleetAuthz,
+    fleetAllowFleetSetupPrivilege,
+  }: { fleetAuthz?: FleetAuthzRouteConfig; fleetAllowFleetSetupPrivilege?: boolean }
+): boolean {
+  if (!checkSecurityEnabled()) {
+    return false;
+  }
 
-interface RouterWrappersSetup {
-  require: {
-    superuser: RouterWrapper;
-    fleetSetupPrivilege: RouterWrapper;
-  };
+  if (fleetAllowFleetSetupPrivilege) {
+    // Temporary while agent call the setup API
+    if (authz.fleet.setup) {
+      return true;
+    }
+  }
+
+  const missingAuthz = [];
+
+  if (fleetAuthz && typeof fleetAuthz.fleet !== 'undefined') {
+    for (const key of fleetAuthz.fleet) {
+      if (!authz.fleet[key as keyof FleetAuthz['fleet']]) {
+        missingAuthz.push(`fleet.${key}`);
+      }
```

Comment:
minor: should we just bail and return early as soon as we find a missing authz rather than creating this array?
```suggestion
      if (!authz.fleet[key as keyof FleetAuthz['fleet']]) {
        return false;
      }
```

Same could be applied below on line 124